### PR TITLE
Consider __toString() implementations in "empty" test [Twig 1.x]

### PR DESCRIPTION
**Pull request by mpdude**
This follows up on #2420:

When you have variables in your views that are actually objects but implement `__toString`, they feel like strings: For example, `{{ something }}` will make use of that to-string-conversion.

What does *not* work is `{% if something is empty %}`, because

> empty checks if a variable is an empty string, an empty array, an empty hash, exactly false, or exactly null  `[http://twig.sensiolabs.org/doc/2.x/tests/empty.html]`

... and obviously `something !== null` in this case.

For template designers, this may be surprising if they don't actually care about the object-or-string difference, they just "use" the variable.

This PR changes the test to consider the return value of the `__toString` method when the value under test is an object, *not* `\Countable` *and* implements `__toString`.

*Yes*, it's a BC break in the edge case of testing (defined) variables that are objects implementing  `__toString` but return the empty string `''`.

### CODE REVIEW

File: lib/Twig/Extension/Core.php [hunk 1]
```diff
@@ -1400,8 +1400,14 @@ function twig_ensure_traversable($seq)
  */
 function twig_test_empty($value)
 {
-    if ($value instanceof Countable) {
-        return 0 == count($value);
+    if (is_object($value)) {
```
Comment (fabpot) [inline lib/Twig/Extension/Core.php]: I think the `is_object()` test is not needed.

File: lib/Twig/Extension/Core.php [hunk 2]
```diff
@@ -1400,8 +1400,14 @@ function twig_ensure_traversable($seq)
...
+        if ($value instanceof Countable) {
+            return 0 == count($value);
+        }
+
+        if (is_callable(array($value, '__toString'))) {
```
Comment (fabpot) [inline lib/Twig/Extension/Core.php]: Same as for the other PR, `method_exists` seems more appropriate in this case.

**Comment by fabpot** [comment]
@mpdude Can you merge this with the other PR as this is really just the same topic? Tests could then be consolidated.

**Comment by mpdude** [comment]
Closing this as changes have been merged into #2420
